docs/design/endpoints.md

```diff
@@ -0,0 +1,292 @@
+     */
+    public open fun resolveBuiltin(value: String)
+
+    /**
+     * Hook for implementers to generate additional custom code that acts on the resolved endpoint, most likely to
+     * delegate further action based on the contents of said endpoint's attributes field.
+     * The AWS SDK, for example, will use this to inspect and handle signing scope information.
+     */
+    public open fun renderPostResolution()
+}
+```
+
+This will generate middleware with the following signature:
+```kotlin
+public class OperationMiddleware(
+    private val config: SmithyService.Config,
+    private val request: OperationRequest,
+    private val provider: EndpointProvider,
+) : ModifyRequestMiddleware {
+    override suspend fun modifyRequest(req: SdkHttpRequest): SdkHttpRequest {
+        /**
+         * Source provider parameters, resolve endpoint, set it on the request.
+         */
+    }
+}
+```
+
+## Example
+
+The following is an example ruleset attached to an arbitrary service definition with the ID "SmithyService" (the
+definition itself is omitted).
+
+The ruleset describes a simple if-else case: if the sole required parameter `ResourceId` contains a specific prefix,
+route this request against a special-case endpoint. Otherwise, fall back to a constant default. The `substring` function
+used in the ruleset is an example of a standard-library function used by endpoint resolvers.
+
+This particular ruleset is kept simple for the purpose of example, naturally, a ruleset can be of arbitrary complexity as
+required by the service.
+
+```json
+{
+  "parameters": {
+    "ResourceId": {
+      "type": "string",
+      "required": "true"
+    }
+  },
+  "rules": [
+    {
+      "documentation": "use a special endpoint for government resources",
+      "type": "endpoint",
+      "conditions": [
+        {
+          "fn": "substring",
+          "argv": [
+            {"ref": "ResourceId"},
+            0,
+            5,
+            false
+          ],
+          "assign": "resourceIdPrefix"
+        },
+        {
+          "fn": "stringEquals",
+          "argv": [
+            {"ref": "resourceIdPrefix"},
+            "gov."
+          ]
+        }
+      ],
+      "endpoint": {
+        "url": "https://gov.api"
+      }
+    },
+    {
+      "documentation": "fallback to global endpoint",
+      "conditions": [],
+      "endpoint": {
+        "url": "https://global.api"
+      }
+    }
+  ]
+}
+```
+*Note*: The signature of the substring method is `substring(s: String, start: Int, end: Int, reverseIndices: false): String`.
+The argv property in the corresponding function call describes how to source the arguments (or input constants directly
+in the case of the indices).
+
+Given a sample operation `GetResource`, with a single operation input `resource` bound to the `ResourceId` provider
+parameter via context traits, this ruleset would roughly generate the following code:
+
+```kotlin
+public data class SmithyServiceEndpointParams(
```

Comment:
to Ian's point below if this is public can we guarantee ordering of parameters will be consistent over time? Otherwise this will break backwards compat one day. 

Perhaps this should follow the builder pattern for operation input/outputs?

Reply:
It _will_ generate a builder for convenience, especially since callers will be able to make these themselves (it should also simplify the codegen). I'll make sure this doc reflects that.

I'm 99% sure we would enforce struct member ordering in params like we would in any other struct, but I'll check.